### ISSUE TITLE
bug(`anvil`): failed transaction with underpriced gas limit shows as success

### ISSUE THREAD

**Issue by Pongch** [foundry-rs/foundry#2998]
I have deployed a contract through low level calls via cast send, the transaction gets mined successfully on Anvil with 'Deployed Contract: <address>' log and 'cast tx' shows the transaction receipt properly

However, the contract does not exist, due to the transaction gas estimate being underpriced

### How to replicate
- make a transaction that requires N amount of gas
- underprice transaction with gasEstimate at N-1

### expected behavior
- show transaction failure and reason on Anvil log or via Cast tx

**Comment by mattsse** [comment]
would appreciate it if you could provide a minimal example.

does cast tx list the transaction as successful or failed?

**Comment by mds1** [comment]
This issue is old, but we actually just reproduced this today. Unfortunately I can't provide our exact case to reproduce, but it was very similar to what @Pongch said. We were simulating a Gnosis Safe transaction and provided all the signatures in calldata (which is why I can't share a repro). Details:

First we used anvil
- It estimated a gas limit of ~117k, but this was insufficient because ~138k gas is is needed
- The transaction receipt showed success, but the full transaction did not actually execute (the expected storage variable change did not occur)
- We looked at the trace with `cast run`. This indicated a top-level revert because of a wrong ecrecover (the recovered address was not a signer on the safe). Only one ecrecover was shown, i.e. it failed on the first signature, and it recovered an unexpected address (the address was not a signer, but all sigs were valid)
- When we lowered the gas limit to ~110k, the results were as expected: failed status, longer trace showing more ecrecovers occuring. 
- When specifying a high enough gas limit, everything works as expected

Then we tried hardhat (via [jackhammer](https://github.com/tcoulter/jackhammer)):
- It chose a better default gas limit so worked out of the box
- We hardcoded the ~117k gas limit anvil used, and saw multiple ecrecovers for all signers before the out of gas (as expected)
- We hardcoded the same ~110k test gas limit, and similarly saw multiple ecrecovers for all signers before the out of gas 

To summarize, there were three issues with anvil:
- Bad gas estimate
- Wrong success value reported
- Wrong trace reported with the 117k gas case

**Comment by zerosnacks** [comment]
Updated the title @mds1, do you think this accurately describes the issue? This was originally tagged as relating to `cast` but this seems related to `anvil` from reading the conversation.

**Comment by grandizzy** [comment]
I am not able to reproduce the initial report (if I correctly understand it), like
- deploy following contract
```Solidity
contract FactoryCounter {
    function createCounter() external {
        new Counter();
    }
}
```
- estimate gas using `cast estimate`
```bash
cast estimate 0x5FbDB2315678afecb367f032d93F642f64180aa3 "createCounter()"
175091
```
- send tx with gas limit 175090 shows status failed
```bash
cast send 0x5FbDB2315678afecb367f032d93F642f64180aa3 "createCounter()" --private-key=0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80 --gas-limit 175090

blockHash               0xe407146383cbe5c7c5ebfcf6808fcd30024186d51dac1482d33bdc4f85e97284
blockNumber             14
contractAddress         
cumulativeGasUsed       173227
effectiveGasPrice       180131337
from                    0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266
gasUsed                 173227
logs                    []
...
status                  0 (failed)
transactionHash         0xb8fea9a94d2b9fcdfa5a0e1e284d5699c53ac18905a6b1530135221319740760
transactionIndex        0
type                    2
blobGasPrice            1
blobGasUsed             
authorizationList       
to                      0x5FbDB2315678afecb367f032d93F642f64180aa3 
```
- same anvil logs shows the revert
```bash
eth_getTransactionReceipt
    Transaction: 0xb8fea9a94d2b9fcdfa5a0e1e284d5699c53ac18905a6b1530135221319740760
    Gas used: 173227
    Error: reverted with: EvmError: Revert

    Block Number: 14
    Block Hash: 0xe407146383cbe5c7c5ebfcf6808fcd30024186d51dac1482d33bdc4f85e97284
    Block Time: "Mon, 25 Nov 2024 10:20:17 +0000"
```

@Pongch is this the right sequence? @mds1 mind to help please with this scenario, would be great to be able to reproduce it and fix it. thanks!

**Comment by mds1** [comment]
Hmm I do not think I'd be able to reproduce this anymore, and I suspect the `createCounter` repro is too simplified to trigger the issue.

Based on the date and gas limit mentioned in my comment, I believe https://etherscan.io/tx/0xea0bb21ac896ddc9fefe9ed2e7ebaf892a7bca3b56ba3d7c272d694e4c83e8e4 is the transaction we were simulating. You could try forking from the block prior and using the same transaction data to repro the bugs

**Comment by jenpaff** [comment]
>Hmm I do not think I'd be able to reproduce this anymore, and I suspect the createCounter repro is too simplified to trigger the issue.

Closing, pls reopen a new issue if you come across it again
